include/ocpp/v201/device_model.hpp

```diff
@@ -168,46 +127,118 @@ class DeviceModel {
+    virtual SetVariableStatusEnum set_value(const Component& component_id, const Variable& variable_id,
+                                            const AttributeEnum& attribute_enum, const std::string& value,
+                                            const bool allow_read_only = false) = 0;
+
     /// \brief Sets the variable_id attribute \p value specified by \p component_id , \p variable_id and \p
     /// attribute_enum for read only variables only. Only works on certain allowed components.
     /// \param component_id
     /// \param variable_id
     /// \param attribute_enum
     /// \param value
     /// \return Result of the requested operation
-    SetVariableStatusEnum set_read_only_value(const Component& component_id, const Variable& variable_id,
-                                              const AttributeEnum& attribute_enum, const std::string& value);
+    virtual SetVariableStatusEnum set_read_only_value(const Component& component_id, const Variable& variable_id,
+                                                      const AttributeEnum& attribute_enum,
+                                                      const std::string& value) = 0;
 
     /// \brief Gets the VariableMetaData for the given \p component_id and \p variable_id
     /// \param component_id
     /// \param variable_id
     /// \return VariableMetaData or std::nullopt if \p component_id or \p variable_id not present
-    std::optional<VariableMetaData> get_variable_meta_data(const Component& component_id, const Variable& variable_id);
+    virtual std::optional<VariableMetaData> get_variable_meta_data(const Component& component_id,
+                                                                   const Variable& variable_id) = 0;
 
     /// \brief Gets the ReportData for the specifed filter \p report_base \p component_variables and \p
     /// component_criteria
     /// \param report_base
     /// \param component_variables
     /// \param component_criteria
     /// \return
-    std::vector<ReportData> get_base_report_data(const ReportBaseEnum& report_base);
+    virtual std::vector<ReportData> get_base_report_data(const ReportBaseEnum& report_base) = 0;
 
     /// \brief Gets the ReportData for the specifed filter \p component_variables and \p
     /// component_criteria
     /// \param report_base
     /// \param component_variables
     /// \param component_criteria
     /// \return
-    std::vector<ReportData>
-    get_custom_report_data(const std::optional<std::vector<ComponentVariable>>& component_variables = std::nullopt,
-                           const std::optional<std::vector<ComponentCriterionEnum>>& component_criteria = std::nullopt);
+    virtual std::vector<ReportData> get_custom_report_data(
+        const std::optional<std::vector<ComponentVariable>>& component_variables = std::nullopt,
+        const std::optional<std::vector<ComponentCriterionEnum>>& component_criteria = std::nullopt) = 0;
```

Comment:
Is it really necessary to transfer an optional `std::vector` instead of just passing an empty vector?

Reply:
I did not really touch or change the functions, I only made an interface for it. So while the comment might be valid I think it is out of scope for this PR so I will not look into it. I will resolve this conversation and also some others which are similar.

Reply:
I see your point.  Still I think, if creating a now an abstract interface for the device model class, it would make sense to rethink some of its interface.  Doing this later on would be at least 3 times the work (abstract interface, implementation and mock).